File: mvp_compat_views/src/dagger/java/com/mandria/android/mvp/basecompatviews/BasePresenterActivity.java
package com.mandria.android.mvp.basecompatviews;

import android.os.Bundle;

import com.mandria.android.mvp.HasPresenter;
import com.mandria.android.mvp.Presenter;
import com.mandria.android.mvp.provider.PresenterProvider;

import javax.inject.Inject;

import androidx.appcompat.app.AppCompatActivity;

/**
 * Base activity for activity which should use a presenter.
 */
public abstract class BasePresenterActivity<P extends Presenter> extends AppCompatActivity implements HasPresenter<P> {

    /**
     * A provider class to handle lifecycle with presenters.
     */
    @Inject
    PresenterProvider mPresenterProvider;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (savedInstanceState != null) {
            mPresenterProvider.onRestoreInstanceState(savedInstanceState.getBundle(PresenterProvider.CONTROLLER_STATE_KEY));
        }

        mPresenterProvider.preparePresenter(this);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBundle(PresenterProvider.CONTROLLER_STATE_KEY, mPresenterProvider.onSaveInstanceState());
    }

    @Override
    protected void onResume() {
        super.onResume();
        mPresenterProvider.attachViewToPresenter(this);
    }

    @Override
    protected void onPause() {
        super.onPause();
        mPresenterProvider.detachViewFromPresenter(isFinishing());
    }

    @Override
    public void finish() {
        mPresenterProvider.detachViewFromPresenter(true);
        super.finish();
    }

    @Override
    public P getPresenter() {
        return mPresenterProvider.getPresenter();
    }

    /**
     * Forces the destruction of a presenter.
     */
    public void destroyPresenter() {
        mPresenterProvider.destroy();
    }
}

File: views_dagger_kotlin/src/main/kotlin/com/mandria/android/mvp/basecompatviews/BasePresenterActivity.kt
package com.mandria.android.mvp.basecompatviews

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.mandria.android.mvp.HasPresenter
import com.mandria.android.mvp.Presenter
import com.mandria.android.mvp.provider.PresenterProvider
import javax.inject.Inject


abstract class BasePresenterActivity<P : Presenter<*>> : AppCompatActivity(), HasPresenter<P> {

    @Inject
    lateinit var presenterProvider: PresenterProvider

    override val presenter: P
        get() = presenterProvider.getPresenter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        savedInstanceState?.let {
            presenterProvider.onRestoreInstanceState(it.getBundle(PresenterProvider.CONTROLLER_STATE_KEY))
        }

        presenterProvider.preparePresenter(this)
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        outState?.putBundle(
            PresenterProvider.CONTROLLER_STATE_KEY,
            presenterProvider.onSaveInstanceState()
        )
    }

    override fun onResume() {
        super.onResume()
        presenterProvider.attachViewToPresenter(this)
    }

    override fun onPause() {
        super.onPause()
        presenterProvider.detachViewFromPresenter(isFinishing)
    }

    override fun finish() {
        presenterProvider.detachViewFromPresenter(true)
        super.finish()
    }

    fun destroyPresenter() {
        presenterProvider.destroy()
    }
}